Ensure `File.get_child()` works with a name, not an ID

Mr0grog opened the pull request:
The documentation for `File.get_child(name)` suggests that it should take a child file’s *name*, not ID. The implementation looks like that’s the intent, too. However, it was actually using the child’s ID instead. This fixes it to use the name.

I didn’t add any tests here since it doesn’t look like there’s any existing tooling to expect or mock the underlying HTTP requests or a live Google test account for it to work against.

It also seems like having types specified in more places would have helped mypy catch this (for example, if `File.id` was typed as `str` and the `raise_if_not_found` argument as a `bool` in `Client.get_file()`, that would have caused mypy to error on the previous code). I can add some types like that if you want.

Mr0grog committed:
Ensure `File.get_child()` works with a name, not an ID

The `.get_child()` method on `File` is documented and designed to take a child file’s *name*, but it previously only worked with an ID. This changes it to take a name instead.

bfontaine commented:
Good catch, thank you!

> I can add some types like that if you want.

Yes, please.

Mr0grog committed:
Add parameter type annotations, fix resulting errors

Mr0grog committed:
Add return types, fix related MyPy failures

Mr0grog commented on drive/client.py:
```diff
@@ -132,21 +132,21 @@ def get_or_create_folder(self, folder_name: str, parent_id: Optional[str] = None
         return self.create_folder(folder_name, parent_id)
 
-    def remove_file(self, file_id: str):
+    def remove_file(self, file_id: str) -> Literal[""]:
```
This is accurate, but seems a little weird! Maybe you want to make this return a bool in the future?

bfontaine replied:
Yep

Mr0grog commented on drive/files.py:
```diff
@@ -95,18 +97,20 @@ def unlink(self):
         return self._client.remove_file(self.id) == ""
 
-    def rename(self, new_name):
+    def rename(self, new_name: str) -> None | Literal[False]:
```
This return type is kinda weird! It is how things behave, though: `False` if no client and `None` otherwise (regardless of whether anything was updated).

bfontaine replied:
Agreed; I’ll first cut a release with the current behavior and then make another one with breaking changes to fix these things.

Mr0grog commented on drive/files.py:
```diff
@@ -95,18 +97,20 @@ def unlink(self):
             return False
+
         m = self._client.update_file(self.id, name=new_name)
         self._update(m)
+        return None
```
MyPy really wanted an explicit return value here.

Mr0grog committed:
Return `False` if no client, like other methods

Mr0grog commented on drive/client.py:
```diff
@@ -500,3 +501,5 @@ def _execute_file_request(self, req):
         if progress:
             print_with_carriage_return('Upload %d%%' %
                                        (100 * progress.progress()))
+
+        return None
```
MyPy really wanted an explicit return here.

Mr0grog commented on drive/files.py:
```diff
@@ -95,29 +97,34 @@ def unlink(self):
         :param new_name:
         :return:
         """
+        if not self._client:
+            return False
+
```
This method couldn’t previously return `False`, but this is the way all other methods handle no `_client`, so I tried to match that behavior here. Otherwise this needs to raise.

bfontaine replied:
Thanks; I’m going to fix this weird behavior at some point and just raise

Mr0grog commented on drive/files.py:
```diff
@@ -196,15 +203,15 @@ def parents(self):
             self.parents_ids = m["parents"]
 
         parents = []
-        for pid in self.parents_ids:
+        for pid in cast(Iterable[str], self.parents_ids):
```
Worth noting: some other type checkers can suss this out automatically (e.g. Pylance). But it seems like MyPy needed this cast. There is similar stuff in one or two other spots.

Mr0grog commented on drive/files.py:
```diff
@@ -223,9 +230,10 @@ def download(self, writer, mime_type=None):
         """
         if not self._client:
             return False
-        return self._client.download(self.id, writer, mime_type=mime_type)
+        self._client.download(self.id, writer, mime_type=mime_type)
+        return None
```
MyPy did not like returning the result of a function that always returns None, so I had to separate the `return` on a separate line (it needs an explicit return because there are multiple return types).

Mr0grog commented on drive/files.py:
```diff
@@ -257,34 +265,37 @@ def get_bytes(self, mime_type=None):
         :return:
         """
         return json.load(self.get_bytes())
 
-    def jsons(self):
+    def jsons(self) -> Generator[Any]:
         """
         Convenient method to download a file as a newline-delimited list of JSON objects. This returns a generator.
         :return:
         """
-        line = True
+        line = b'1'
```
This keeps the typing simple. Otherwise this variable needs to be annotated as `bool | bytes`.

Mr0grog commented on drive/files.py:
```diff
@@ -294,22 +305,22 @@ def to_dict(self):
 
         return d
 
-    def __str__(self):
-        return self.name
+    def __str__(self) -> str:
+        return self.name or ''
```
I think this is the only spot where I really changed behavior (other than returning false one place where `self._client` is missing). Not sure what you’d like to see here when `name` is missing, but the previously implementation made MyPy unhappy since this should always be returning a string.

bfontaine replied:
Thank you; it’s fine like you did for now

Mr0grog reviewed (commented):
Sorry this took me a little time to get to.

I went through and added argument and return types for all functions (and fixed some where the current types seemed incorrect). Some stuff here is a little weird!

Some notes:

- `Client.get_shared_directory()` is typed as returning `Optional[File]` which is true as far as a type checker is concerned, but in practice always returns `File` (or raises). I considered adding a `cast` to make that possible, but it seemed really error-prone if the default for `raise_if_not_found` in `get_shard_file()` changes.

- In general, where things raise vs. return `None` vs. return `False` feels a little idiosyncratic. When I had to add some checks to satisfy MyPy, I did my best to follow the existing style of similar methods, but I don’t feel confident the choices I made were obviously correct.

- `raise_if...` arguments in some of the client functions make the types more complicated than they might otherwise be.

- Similarly, `_execute_file_request()` forces a lot of `cast()` calls. It might be a good idea to refactor that into separate functions that make clearer assertions about what you should be getting back, but that’s way outside the scope of this. I was also surprised to notice it returns `None` if it’s marked as resumable (doesn’t try to parse the final result). I don’t *think* I did any dangerous `cast` calls that are hiding possible `None` values, but you might want to check that over.

bfontaine commented on drive/client.py:
```diff
@@ -92,30 +92,30 @@ def _serialize_query_value(value):
-        http = authorize_credentials(credentials_path)
-        self.service = discovery.build('drive', 'v3', http=http)
-        self.download_retries_count = download_retries_count
+    def __init__(self, credentials_path: Optional[str] = None, *, download_retries_count: int = 5) -> None:
```
this `: int` should not be necessary since `5` is already an `int`.

Mr0grog replied:
That’s not really true — if you don’t explicitly define the type this way, it will be treated as `Any` (at least in current versions of Python and MyPy).

For example, the MyPy says the following code is OK if you take away the `: int` annotation here, but will error (which is what you want) if you keep it:

```py
Client('edgi-testing-service.json', download_retries_count='5')
```

I can change these back if you want, but it kind of works against the reason I added all these type hints.

bfontaine replied:
Argh you’re right; I mistakenly thought that because of IntelliJ’s typing, which does show an error on `download_retries_count='5'` even without `: int`.

bfontaine commented on drive/client.py:
```diff
@@ -92,30 +92,30 @@ def _serialize_query_value(value):
+    def __init__(self, credentials_path: Optional[str] = None, *, download_retries_count: int = 5) -> None:
+        http: httplib2.Http = authorize_credentials(credentials_path)
+        self.service: Any = discovery.build('drive', 'v3', http=http)
+        self.download_retries_count: int = download_retries_count
```
This `: int` is not needed because `download_retries_count` is already typed as an `int`

Mr0grog replied:
It is not — this is typing the instance variable on `self`, which is treated as `Any` if you don’t do this. For code outside this method, `self.download_retries_count` is essentially of an unknown type unless you set it explicitly here. (Another option is to define the types directly in the class instead of in `__init__`:

```py
class Client:
    download_retries_count: int
```

Happy to switch to that style if you’d prefer it. But one way or the other, these instance variables do need their types set — they are not inferred from the type of the local variable they are being set from. (I don’t think Pyright will infer this either, but maybe. it does do a lot more intelligent inference than MyPy does.)

bfontaine commented on drive/client.py:
```diff
@@ -92,30 +92,30 @@ def _serialize_query_value(value):
         return self.service.permissions()
 
-    def create_folder(self, name: str, parent_id: Optional[str] = None):
+    def create_folder(self, name: str, parent_id: Optional[str] = None) -> File:
```
Isn’t this annotation redundant with the `return cast(...)`?

Mr0grog replied:
No, not in MyPy (I *think* Pyright can infer this). Try it out:

```python
from typing import assert_type

client = Client()
assert_type(client.create_folder('whatever', 'abc123'), File)
```

Without the explicit return type, this will error because it thinks the type is `Any`.

bfontaine commented on drive/client.py:
```diff
@@ -202,11 +202,11 @@ def files_shared_with_me(self) -> List[File]:
     def get_shared_file(self, name: str,
                         is_directory: Optional[bool] = None,
-                        raise_if_not_found=True) -> Optional[File]:
+                        raise_if_not_found: bool = True) -> Optional[File]:
```
This `: bool` is not needed because `True` is already a `bool`.

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -243,7 +243,7 @@ def list_files(self,
                    mimetype: Optional[str] = None,
                    parents_in: Optional[str] = None,
-                   n=100):
+                   n: int = 100) -> List[File]:
```
`: int` is not needed because `100` is already an `int`. `List[File]` is redundant with the `return cast(...)`

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -261,14 +261,14 @@ def list_files(self,
-                    media=None,
-                    force=False):
+                    media: Optional[MediaUpload] = None,
+                    force: bool = False) -> Optional[File]:
```
`: bool` is not needed because `False` is already a `bool`. I think `Optional[File]` is redundant because Mypy can already see that the function returns either `None` either `cast(File, ...)`

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -351,7 +351,7 @@ def download_file(self, file_id: str, path: str, mime_type: Optional[str] = None
             self.download(file_id, f, mime_type=mime_type)
 
-    def download_excel_workbook(self, file_id: str, read_only=False):
+    def download_excel_workbook(self, file_id: str, read_only: bool = False) -> openpyxl.Workbook:
```
```suggestion
    def download_excel_workbook(self, file_id: str, read_only = False) -> openpyxl.Workbook:
```
see other similar comments above

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -365,11 +365,11 @@ def download_excel_workbook(self, file_id: str, read_only=False):
                original_mime_type: Optional[str] = None,
-               update_existing=False,
-               resumable=False):
+               update_existing: bool = False,
```
```suggestion
               update_existing = False,
```

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -365,11 +365,11 @@ def download_excel_workbook(self, file_id: str, read_only=False):
-               update_existing=False,
-               resumable=False):
+               update_existing: bool = False,
+               resumable: bool = False) -> File:
```
```suggestion
               resumable = False) -> File:
```

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -403,14 +403,15 @@ def upload(self, parent_id: Union[str, File], name: str,
                     mime_type: Optional[str] = None,
                     original_mime_type: Optional[str] = None,
-                    update_existing=False):
+                    update_existing: bool = False) -> File:
```
```suggestion
                    update_existing = False) -> File:
```

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -433,8 +434,8 @@ def upload_excel_workbook(self,
                               workbook: openpyxl.Workbook,
-                              as_spreadsheet=True,
-                              update_existing=False):
+                              as_spreadsheet: bool = True,
```
```suggestion
                              as_spreadsheet = True,
```

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/client.py:
```diff
@@ -433,8 +434,8 @@ def upload_excel_workbook(self,
-                              as_spreadsheet=True,
-                              update_existing=False):
+                              as_spreadsheet: bool = True,
+                              update_existing: bool = False) -> File:
```
```suggestion
                              update_existing = False) -> File:
```

Mr0grog replied:
See above: https://github.com/bfontaine/drive/pull/13#discussion_r2015341970

bfontaine commented on drive/files.py:
```diff
@@ -2,20 +2,21 @@
     """
 
-    def __init__(self, attrs, client: Optional["drive.Client"] = None):
+    def __init__(self, attrs: Union["File", dict[Any, Any]], client: Optional["drive.Client"] = None) -> None:
```
Can’t we use `dict[str, Any]` instead of `dict[Any, Any]`?

Mr0grog replied:
Sure. I did not know what the intended usage was — there’s no documentation that describes what the expected input here is, so it wasn’t clear to me if you only wanted this to accept dicts with string keys or any dict that happens to have the string keys we care about (since there could be a mix of key types).

bfontaine commented on drive/sheets.py:
```diff
@@ -12,8 +12,8 @@
 def _auto_retry(fn: Callable[[], T], *,
-                max_retries=2,
-                sleep_for=4) -> T:
+                max_retries: int = 2,
```
```suggestion
                max_retries = 2,
```
`2` is already an `int`

bfontaine commented on drive/sheets.py:
```diff
@@ -12,8 +12,8 @@
-                max_retries=2,
-                sleep_for=4) -> T:
+                max_retries: int = 2,
+                sleep_for: int = 4) -> T:
```
```suggestion
                sleep_for = 4) -> T:
```
idem

bfontaine commented on drive/sheets.py:
```diff
@@ -29,13 +29,13 @@ def _auto_retry(fn: Callable[[], T], *,
 
     def get_sheet_range(self, sheet_id: str, sheet_tab: str, cell_range: str,
-                        *, max_retries=2):
+                        *, max_retries: int = 2):
```
```suggestion
                        *, max_retries = 2):
```

bfontaine commented on drive/sheets.py:
```diff
@@ -52,9 +52,9 @@ def get_sheet_range(self, sheet_id: str, sheet_tab: str, cell_range: str,
-                         offset=0,
-                         batch_size=400,
-                         sleep_for=0.5,
+                         offset: int = 0,
```
```suggestion
                         offset = 0,
```

bfontaine commented on drive/sheets.py:
```diff
@@ -52,9 +52,9 @@ def get_sheet_range(self, sheet_id: str, sheet_tab: str, cell_range: str,
-                         batch_size=400,
-                         sleep_for=0.5,
+                         offset: int = 0,
+                         batch_size: int = 400,
```
```suggestion
                         batch_size = 400,
```

bfontaine commented on drive/sheets.py:
```diff
@@ -52,9 +52,9 @@ def get_sheet_range(self, sheet_id: str, sheet_tab: str, cell_range: str,
-                         sleep_for=0.5,
+                         offset: int = 0,
+                         batch_size: int = 400,
+                         sleep_for: float = 0.5,
```
```suggestion
                         sleep_for = 0.5,
```
`0.5` is already a `float`

bfontaine commented:
Thank you! I didn’t expect you to go all over the code :sweat_smile: 

Yes I don’t like either that mix of `None`/`False`/exceptions; I’m going to make a release with your changes to stay backward-compatible, and then I’ll do a new major release that uses exceptions for everytime there is unexpected behavior.

Mr0grog committed:
Minor bits missed in the previous commit

I had these locally, so MyPy was passing, but somehow missed committing them.

Mr0grog committed:
Fix annotations not compatible with Python 3.9

Mr0grog committed:
Upgrade MyPy to current release for better typing support

Mr0grog commented:
Just noticed the test failure — sorry I somehow missed a couple lines that were in my local copy yesterday. I also realized I was working in Python 3.12 and used some annotation styles that aren’t supported in 3.9, so I went back and fixed that.

Finally, I went ahead and updated MyPy here, since it is past an actual 1.0 now, and I used some typings that pre-1.0 MyPy does not support (for generators).

Mr0grog committed:
Python 3.9 requires all 3 args for Generator

bfontaine commented:
Thank you! I merged and release v0.4.5.

Now I’m fixing my Python code everywhere to add explicit type hints to arguments with default values :joy:
